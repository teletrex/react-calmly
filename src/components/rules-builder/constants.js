/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




// comment needed for translation parsing
// t('Or')
// t('And')
// t('And (changes all in level)')
// t('And (adds bracket)')
// t('Or (changes all in level)')
// t('Or (adds bracket)')

export const CONJUNCTION_OR = Object.freeze({
  id: 'OR',
  label: 'Or',
});

export const CONJUNCTION_AND = Object.freeze({
  id: 'AND',
  label: 'And',
});

export const CONJUNCTION_AND_ALL = Object.freeze({
  id: 'AND_ALL',
  value: 'AND',
  label: 'And (changes all in level)',
  changeAll: true,
});

export const CONJUNCTION_AND_BRACKETS = Object.freeze({
  id: 'AND_BRACKETS',
  value: 'AND',
  label: 'And (adds bracket)',
  changeAll: false,
});

export const CONJUNCTION_OR_ALL = Object.freeze({
  id: 'OR_ALL',
  value: 'OR',
  label: 'Or (changes all in level)',
  changeAll: true,
});

export const CONJUNCTION_OR_BRACKETS = Object.freeze({
  id: 'OR_BRACKETS',
  value: 'OR',
  label: 'Or (adds bracket)',
  changeAll: false,
});

export const RULE_TYPES = {
  custom: 'custom',
  date: 'date',
  number: 'number',
  select: 'select',
  string: 'string',
};

export const RULE_INPUT_TYPES = {
  // Common:
  RANGE: 'range',
  NONE: 'none',
  SINGLE_VALUE: 'singleValue',
  MULTIPLE_VALUES: 'multipleValues',

  // Date related:
  PERIOD_NUMBER: 'periodNumber',
  WHEN_RELATION: 'whenRelation',
  WITHIN_PERIOD: 'withinPeriod',
  PERIOD_RANGE: 'periodRange',
};

export const CONJUNCTIONS = {
  CONJUNCTION_OR,
  CONJUNCTION_AND,
  CONJUNCTION_AND_ALL,
  CONJUNCTION_AND_BRACKETS,
  CONJUNCTION_OR_ALL,
  CONJUNCTION_OR_BRACKETS,
};
