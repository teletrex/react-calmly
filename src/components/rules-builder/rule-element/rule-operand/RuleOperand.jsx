/* <LICENSE>
*
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
*
* </LICENSE>
*  */




import React, { useCallback, useContext } from 'react';
import PropTypes from 'prop-types';
import isNull from 'lodash/isNull';
import noop from 'lodash/noop';
import settings from '../../../../settings';

import {
  TextIndentLess,
  TextIndentMore,
  SubtractAlt,
  Draggable,
  ErrorFilled,
} from '@carbon/icons-react';

import classNames from 'classnames';

import RulesBuilderContext from '../../RulesBuilderContext';
import { useTranslation } from '../../../../translation';
import RuleContainer from '../../rules/components/RuleContainer';

import CriteriaDropdown from './criteria-dropdown';
import RuleElementMenu from './rule-element-menu';

const { prefix } = settings;
const defaultClassName = `${prefix}--rule-element--item`;

const RuleOperand = ({
  hideDeleteBtn,
  isDragging,
  isOnlyElement,
  isFirst,
  isLast,
  listeners,
  onCreateGroup,
  onDeleteRule,
  onDuplicate,
  onMoveDown,
  onMoveUp,
  onRemoveBracket,
  onCriteriaChange,
  onRuleChange,
  onTimeframeChange,
  rule,
}) => {
  const { t } = useTranslation();
  const {
    maxDepth,
    isLight,
    isReadOnly,
    customLabels: { criteriaDropdownLabel = '' },
  } = useContext(RulesBuilderContext);

  const { data: ruleComponentData = {}, errorMessage } = rule;
  const { RuleComponent, ruleConfig, ruleData, ruleTimeframe } = ruleComponentData || {};

  const timeframe = isNull(ruleTimeframe)
    ? {}
    : {
        data: ruleTimeframe,
        onChange: onTimeframeChange,
      };

  const handleRemoveBracket = useCallback(() => onRemoveBracket(rule.id), [
    onRemoveBracket,
    rule.id,
  ]);

  const handleDeleteRule = useCallback(() => onDeleteRule(rule.id), [onDeleteRule, rule.id]);

  const handleCreateGroup = useCallback(() => {
    onCreateGroup(rule.id);
  }, [onCreateGroup, rule.id]);

  const handleCriteriaChange = useCallback(criteriaKey => onCriteriaChange(rule.id, criteriaKey), [
    onCriteriaChange,
    rule.id,
  ]);

  const handleDuplicate = useCallback(() => onDuplicate(rule.id), [onDuplicate, rule.id]);

  const handleMoveUp = useCallback(() => onMoveUp(rule.id), [onMoveUp, rule.id]);

  const handleMoveDown = useCallback(() => onMoveDown(rule.id), [onMoveDown, rule.id]);

  const renderDragHandler = () => (
    <button
      {...listeners}
      aria-label={t('Draggable item')}
      className={`${defaultClassName}--draggable-btn`}
      type="button"
    >
      <Draggable size={16} className={`${defaultClassName}--btn--icon`} />
    </button>
  );

  const renderCriteriaDropdown = () => {
    if (ruleConfig?.disableCriteriaReselect) {
      return null;
    }

    return (
      <CriteriaDropdown
        className={`${defaultClassName}--criteria-dropdown`}
        label={criteriaDropdownLabel}
        onChange={handleCriteriaChange}
        selectedCriteriaKey={rule.criteriaKey}
      />
    );
  };

  return (
    <>
      <div
        className={classNames(`${defaultClassName}`, {
          [`${defaultClassName}--error`]: !!errorMessage,
          [`${defaultClassName}--readonly`]: isReadOnly,
          [`${defaultClassName}--dragging`]: isDragging,
          [`${defaultClassName}--light`]: isLight,
        })}
      >
        {renderDragHandler()}
        <div className={`${defaultClassName}--content`}>
          {renderCriteriaDropdown()}
          {RuleComponent && ruleConfig && (
            <RuleContainer
              config={ruleConfig}
              data={ruleData}
              onRuleChange={onRuleChange}
              ruleComponent={RuleComponent}
              timeframe={timeframe}
            />
          )}
        </div>

        <div className={`${defaultClassName}--controls`}>
          {isOnlyElement && rule.level > 1 && (
            <button
              aria-label={t('Remove bracket')}
              className={classNames(
                `${defaultClassName}--overflow-menu`,
                `${prefix}--overflow-menu`
              )}
              onClick={handleRemoveBracket}
              type="button"
            >
              <TextIndentLess size={16} className={`${defaultClassName}--btn--icon`} />
            </button>
          )}
          {!isOnlyElement && rule.level < maxDepth && (
            <button
              aria-label={t('Create group')}
              className={classNames(
                `${defaultClassName}--overflow-menu`,
                `${prefix}--overflow-menu`
              )}
              onClick={handleCreateGroup}
              type="button"
            >
              <TextIndentMore size={16} className={`${defaultClassName}--btn--icon`} />
            </button>
          )}
          {!hideDeleteBtn && (
            <button
              aria-label={t('Delete rule')}
              className={classNames(
                `${defaultClassName}--overflow-menu`,
                `${prefix}--overflow-menu`
              )}
              data-testid="delete-rule"
              onClick={handleDeleteRule}
              type="button"
            >
              <SubtractAlt size={16} className={`${defaultClassName}--btn--icon`} />
            </button>
          )}
          <RuleElementMenu
            className={`${defaultClassName}--overflow-menu`}
            duplicateDisabled={!rule.criteriaKey}
            moveDownDisabled={isLast && rule.level === 1}
            moveUpDisabled={isFirst && rule.level === 1}
            onDuplicate={handleDuplicate}
            onMoveDown={handleMoveDown}
            onMoveUp={handleMoveUp}
          />
          {errorMessage && <ErrorFilled size={16} className={`${defaultClassName}--error-icon`} />}
        </div>
      </div>
      {errorMessage && <p className={`${defaultClassName}--error-text`}>{errorMessage}</p>}
    </>
  );
};

const RulePropType = {
  criteriaKey: PropTypes.string,
  id: PropTypes.string,
  level: PropTypes.number.isRequired,
};

RulePropType.data = PropTypes.oneOfType([
  PropTypes.object,
  PropTypes.arrayOf(PropTypes.shape(RulePropType)),
]);

RuleOperand.propTypes = {
  hideDeleteBtn: PropTypes.bool,
  isDragging: PropTypes.bool,
  isFirst: PropTypes.bool,
  isLast: PropTypes.bool,
  isOnlyElement: PropTypes.bool,
  listeners: PropTypes.shape({ onKeyDown: PropTypes.func, onPointerDown: PropTypes.func }),
  onCreateGroup: PropTypes.func,
  onCriteriaChange: PropTypes.func,
  onDeleteRule: PropTypes.func,
  onDuplicate: PropTypes.func,
  onMoveDown: PropTypes.func,
  onMoveUp: PropTypes.func,
  onRemoveBracket: PropTypes.func,
  onRuleChange: PropTypes.func,
  onTimeframeChange: PropTypes.func,
  rule: PropTypes.shape(RulePropType).isRequired,
};

RuleOperand.defaultProps = {
  hideDeleteBtn: false,
  isDragging: false,
  isFirst: false,
  isLast: false,
  isOnlyElement: false,
  listeners: {},
  onCreateGroup: noop,
  onCriteriaChange: noop,
  onDeleteRule: noop,
  onDuplicate: noop,
  onMoveDown: noop,
  onMoveUp: noop,
  onRemoveBracket: noop,
  onRuleChange: noop,
  onTimeframeChange: noop,
};

export default RuleOperand;
