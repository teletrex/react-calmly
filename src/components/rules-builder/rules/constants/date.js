/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export const DATE_PATTERN = 'd{1,2}/d{1,2}/d{4}';
