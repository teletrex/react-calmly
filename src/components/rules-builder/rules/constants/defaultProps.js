/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export const DEFAULT_PROPS = {
  INPUT_FIELD_VALIDATION: [],
  OPERATORS: [],
  TIMEFRAME: {},
};
