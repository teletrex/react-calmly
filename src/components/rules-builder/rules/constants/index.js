/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export * from './date';
export * from './defaultProps';
export * from './periods';
export * from './propTypes';
export * from './ruleOperators';
export * from './validationMessages';
export * from './values';
export * from './whenRelations';
