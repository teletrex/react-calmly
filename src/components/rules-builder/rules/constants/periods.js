/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export const PERIODS = [
  {
    label: 'days',
    value: 'DAYS',
  },
  {
    label: 'weeks',
    value: 'WEEKS',
  },
  {
    label: 'months',
    value: 'MONTHS',
  },
];
