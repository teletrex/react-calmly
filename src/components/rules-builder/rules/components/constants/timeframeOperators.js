/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import { RULE_INPUT_TYPES } from '../../../constants';

const { NONE, RANGE, SINGLE_VALUE, WHEN_RELATION, WITHIN_PERIOD } = RULE_INPUT_TYPES;

export const TIMEFRAME_OPERATORS = [
  // Within:
  {
    inputType: WITHIN_PERIOD,
    label: 'within the last',
    value: 'WITHIN_THE_LAST',
  },

  // Today:
  {
    inputType: NONE,
    label: 'on today',
    value: 'ON_TODAY',
  },
  {
    inputType: NONE,
    label: 'before today',
    value: 'BEFORE_TODAY',
  },

  // Any time:
  {
    inputType: NONE,
    label: 'at any time',
    value: 'AT_ANY_TIME',
  },

  // Date:
  {
    inputType: SINGLE_VALUE,
    label: 'on',
    value: 'ON',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'after',
    value: 'AFTER',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'before',
    value: 'BEFORE',
  },

  // Date range:
  {
    inputType: RANGE,
    label: 'between',
    value: 'BETWEEN',
  },

  // When relation:
  {
    inputType: WHEN_RELATION,
    label: 'exactly',
    value: 'EXACTLY',
  },
  {
    inputType: WHEN_RELATION,
    label: 'more than',
    value: 'MORE_THAN',
  },
  {
    inputType: WHEN_RELATION,
    label: 'less than',
    value: 'LESS_THAN',
  },
];
