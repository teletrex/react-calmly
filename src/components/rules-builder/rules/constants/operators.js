/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import { RULE_INPUT_TYPES } from '../../constants';

const {
  MULTIPLE_VALUES,
  NONE,
  PERIOD_NUMBER,
  PERIOD_RANGE,
  RANGE,
  SINGLE_VALUE,
  WHEN_RELATION,
  WITHIN_PERIOD,
} = RULE_INPUT_TYPES;

// Common operators
export const BLANK_OPERATORS = [
  {
    inputType: NONE,
    label: 'is blank',
    value: 'IS_BLANK',
  },
  {
    inputType: NONE,
    label: 'is not blank',
    value: 'IS_NOT_BLANK',
  },
];

export const EXACT_OPERATORS = [
  {
    inputType: WHEN_RELATION,
    label: 'is exactly',
    value: 'IS_EXACTLY',
  },
  {
    inputType: WHEN_RELATION,
    label: 'is not exactly',
    value: 'IS_NOT_EXACTLY',
  },
];

export const IS_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'is',
    value: 'IS',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'is not',
    value: 'IS_NOT',
  },
];

export const HAS_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'has',
    value: 'HAS',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'has not',
    value: 'HAS_NOT',
  },
];

export const IS_ONE_OF_OPERATORS = [
  {
    inputType: MULTIPLE_VALUES,
    label: 'is one of the following',
    value: 'IS_ONE_OF_THE_FOLLOWING',
  },
  {
    inputType: MULTIPLE_VALUES,
    label: 'is not one of the following',
    value: 'IS_NOT_ONE_OF_THE_FOLLOWING',
  },
];

export const CONTAIN_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'contains',
    value: 'CONTAINS',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'does not contain',
    value: 'DOES_NOT_CONTAIN',
  },
];

export const CONTAIN_ONE_OF_OPERATORS = [
  {
    inputType: MULTIPLE_VALUES,
    label: 'contains one of the following',
    value: 'CONTAINS_ONE_OF_THE_FOLLOWING',
  },
  {
    inputType: MULTIPLE_VALUES,
    label: 'does not contain one of the following',
    value: 'DOES_NOT_CONTAIN_ONE_OF_THE_FOLLOWING',
  },
];

export const RANGE_OPERATORS = [
  {
    inputType: RANGE,
    label: 'is between',
    value: 'IS_BETWEEN',
  },
];

// String related operators
export const STARTS_WITH_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'starts with',
    value: 'STARTS_WITH',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'does not start with',
    value: 'DOES_NOT_START_WITH',
  },
];

export const STARTS_WITH_ONE_OF_OPERATORS = [
  {
    inputType: MULTIPLE_VALUES,
    label: 'starts with one of the following',
    value: 'STARTS_WITH_ONE_OF_THE_FOLLOWING',
  },
  {
    inputType: MULTIPLE_VALUES,
    label: 'does not start with one of the following',
    value: 'DOES_NOT_START_WITH_ONE_OF_THE_FOLLOWING',
  },
];

export const ENDS_WITH_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'ends with',
    value: 'ENDS_WITH',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'does not end with',
    value: 'DOES_NOT_END_WITH',
  },
];

export const ENDS_WITH_ONE_OF_OPERATORS = [
  {
    inputType: MULTIPLE_VALUES,
    label: 'ends with one of the following',
    value: 'ENDS_WITH_ONE_OF_THE_FOLLOWING',
  },
  {
    inputType: MULTIPLE_VALUES,
    label: 'does not end with one of the following',
    value: 'DOES_NOT_END_WITH_ONE_OF_THE_FOLLOWING',
  },
];

// Number related operators
export const LESS_THAN_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'is less than',
    value: 'IS_LESS',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'is less than or equal to',
    value: 'IS_LESS_OR_EQUAL',
  },
];

export const GREATER_THAN_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'is greater than',
    value: 'IS_GREATER',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'is greater than or equal to',
    value: 'IS_GREATER_OR_EQUAL',
  },
];

// Date related operators
export const TODAY_OPERATORS = [
  {
    inputType: NONE,
    label: 'is today',
    value: 'IS_TODAY',
  },
  {
    inputType: NONE,
    label: 'is after today',
    value: 'IS_AFTER_TODAY',
  },
  {
    inputType: NONE,
    label: 'is not after today',
    value: 'IS_NOT_AFTER_TODAY',
  },
  {
    inputType: NONE,
    label: 'is before today',
    value: 'IS_BEFORE_TODAY',
  },
  {
    inputType: NONE,
    label: 'is not before today',
    value: 'IS_NOT_BEFORE_TODAY',
  },
  {
    inputType: NONE,
    label: 'anniversary is today',
    value: 'ANNIVERSARY_IS_TODAY',
  },
];

export const ANNIVERSARY_OPERATORS = [
  {
    inputType: WHEN_RELATION,
    label: 'anniversary is',
    value: 'ANNIVERSARY_IS',
  },
];

export const WITHIN_OPERATORS = [
  {
    inputType: WITHIN_PERIOD,
    label: 'is within the last',
    value: 'IS_WITHIN_THE_LAST',
  },
  {
    inputType: WITHIN_PERIOD,
    label: 'is not within the last',
    value: 'IS_NOT_WITHIN_THE_LAST',
  },
  {
    inputType: WITHIN_PERIOD,
    label: 'is within the next',
    value: 'IS_WITHIN_THE_NEXT',
  },
  {
    inputType: WITHIN_PERIOD,
    label: 'is not within the next',
    value: 'IS_NOT_WITHIN_THE_NEXT',
  },
];

export const WITHIN_RANGE_OPERATORS = [
  {
    inputType: PERIOD_RANGE,
    label: 'is within the range of the last',
    value: 'IS_WITHIN_THE_RANGE_OF_THE_LAST',
  },
];

export const PERIOD_EQUAL_OPERATORS = [
  {
    inputType: PERIOD_NUMBER,
    label: 'year is',
    value: 'YEAR_IS',
  },
  {
    inputType: PERIOD_NUMBER,
    label: 'month is',
    value: 'MONTH_IS',
  },
  {
    inputType: PERIOD_NUMBER,
    label: 'day is',
    value: 'DAY_IS',
  },
];

export const BEFORE_AFTER_DATE_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'is after',
    value: 'IS_AFTER',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'is before',
    value: 'IS_BEFORE',
  }
]

export const DATA_AGGREGATION_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'by',
    value: 'BY',
  },

];
