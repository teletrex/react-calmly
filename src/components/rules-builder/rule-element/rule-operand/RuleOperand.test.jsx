/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React from 'react';
import { render } from '@testing-library/react';

import { CONJUNCTION_AND } from '../../constants';

import RuleOperand from './RuleOperand';

describe('RuleOperand', () => {
  const defaultRule = {
    id: '1',
    key: 'AGE',
    conjunction: CONJUNCTION_AND.id,
    data: {},
    level: 1,
  };

  it('should render correctly', () => {
    const props = { rule: defaultRule };

    const { container } = render(<RuleOperand {...props} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should allow to reselect criteria', () => {
    const props = { rule: defaultRule };

    const { getByText } = render(<RuleOperand {...props} />);

    expect(getByText('Select criteria')).toBeInTheDocument();
  });

  it('should not render ungroup button', () => {
    const props = { rule: { ...defaultRule }, isOnlyElement: true };

    const { queryByLabelText } = render(<RuleOperand {...props} />);

    expect(queryByLabelText('Remove bracket')).not.toBeInTheDocument();
  });

  it('should render ungroup button', () => {
    const props = { rule: { ...defaultRule, level: 2 }, isOnlyElement: true };

    const { getByLabelText } = render(<RuleOperand {...props} />);

    expect(getByLabelText('Remove bracket')).toBeInTheDocument();
  });
});
