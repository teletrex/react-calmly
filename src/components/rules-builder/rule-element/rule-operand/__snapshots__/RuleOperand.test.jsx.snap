// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RuleOperand should render correctly 1`] = `
<div
  class="bx--rule-element--item"
>
  <button
    aria-label="Draggable item"
    class="bx--rule-element--item--draggable-btn"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="bx--rule-element--item--btn--icon"
      focusable="false"
      height="16"
      preserveAspectRatio="xMidYMid meet"
      style="will-change: transform;"
      viewBox="0 0 16 16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M5 3h2v2H5zm4 0h2v2H9zM5 7h2v2H5zm4 0h2v2H9zm-4 4h2v2H5zm4 0h2v2H9z"
      />
    </svg>
  </button>
  <div
    class="bx--rule-element--item--content"
  >
    <div
      class="bx-proposal-dropdown-wrapper bx-proposal-dropdown-wrapper--inline"
    >
      <div
        class="bx--rule-element--item--criteria-dropdown bx-proposal-dropdown"
        role="listbox"
      >
        <span>
          <div
            aria-label="Select criteria"
            class="bx-proposal-dropdown__field"
            role="button"
            tabindex="0"
          >
            <span
              class="bx-proposal-dropdown__field__label"
            >
              Select criteria
            </span>
            <div
              class="bx-proposal-dropdown__field__menu-icon"
            >
              <svg
                aria-hidden="true"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                style="will-change: transform;"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 11L3 6l.7-.7L8 9.6l4.3-4.3.7.7z"
                />
              </svg>
            </div>
          </div>
        </span>
      </div>
    </div>
  </div>
  <div
    class="bx--rule-element--item--controls"
  >
    <button
      aria-label="Create group"
      class="bx--rule-element--item--overflow-menu bx--overflow-menu"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="bx--rule-element--item--btn--icon"
        focusable="false"
        height="16"
        preserveAspectRatio="xMidYMid meet"
        style="will-change: transform;"
        viewBox="0 0 32 32"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20 6h10v2H20zm0 6h7v2h-7zm0 6h10v2H20zm0 6h7v2h-7zM14 4h2v24h-2zM3.85 22.5l-1.41-1.38L8.73 15 2.44 8.88 3.85 7.5l7.71 7.5-7.71 7.5z"
        />
      </svg>
    </button>
    <button
      aria-label="Delete rule"
      class="bx--rule-element--item--overflow-menu bx--overflow-menu"
      data-testid="delete-rule"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="bx--rule-element--item--btn--icon"
        focusable="false"
        height="16"
        preserveAspectRatio="xMidYMid meet"
        style="will-change: transform;"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 2c3.3 0 6 2.7 6 6s-2.7 6-6 6-6-2.7-6-6 2.7-6 6-6m0-1C4.1 1 1 4.1 1 8s3.1 7 7 7 7-3.1 7-7-3.1-7-7-7z"
        />
        <path
          d="M5 7.5h6v1H5z"
        />
      </svg>
    </button>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="open and close list of options"
      class="bx--rule-element--item--overflow-menu bx--overflow-menu"
      data-testid="overflow-menu"
      direction="bottom"
      group-id="default_overflow_menus"
      i18n="[object Object]"
      type="button"
    >
      <svg
        aria-label="open and close list of options"
        class="bx--overflow-menu__icon"
        focusable="false"
        height="16"
        preserveAspectRatio="xMidYMid meet"
        role="img"
        style="will-change: transform;"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="8"
          cy="3"
          r="1"
        />
        <circle
          cx="8"
          cy="8"
          r="1"
        />
        <circle
          cx="8"
          cy="13"
          r="1"
        />
        <title>
          open and close list of options
        </title>
      </svg>
    </button>
  </div>
</div>
`;
